doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,297 @@
+
+This is a common approach for advanced search applications.
+
+The ranking of search results on websites is frequently determined by
+machine learning models. For example, Airbnb, Yelp, Wikipedia, Bloomberg,
+and commercial search engines.
+
+There is a good explanation on what LTR is here:
+
+- https://elasticsearch-learning-to-rank.readthedocs.io/en/latest/core-concepts.html
+- https://opensourceconnections.com/blog/2017/02/24/what-is-learning-to-rank/
+
+### How does Learning To Rank work?
+
+It's easiest to explain with an example:
+
+Say you searched for 'Harry Potter'. Search API might return the following results:
+
+1. HMRC Sign in
+2. Harry Potter World
+3. Pottery barn
+4. Harry Potter
+
+These are not good results.
+
+To assess search quality, we can assign relevancy judgements to the
+results of a query in the range 0-3, with 0 being irrelevant and 3
+being perfect. For example, for these results:
+
+```
+query         rating    document
+harry potter  0         HMRC Sign in
+harry potter  2         Harry Potter World
+harry potter  0         Pottery barn
+harry potter  3         Harry Potter
+```
+
+We can then use a ranking quality metric, such as [normalised
+discounted cumulative gain (nDCG)](https://en.wikipedia.org/wiki/Discounted_cumulative_gain#Normalized_DCG) to assign a score to
+this list of results.
+
+We can then use nDCG to tell us how good the results are.
+We might get a score of 0.3 for this result set (from 0 to 1).
+
+We could then investigate why the results are bad. Perhaps there is a
+best bet. Or popularity might be pushing HMRC docs up. We can fix these
+things, then check the new nDCG score, and we might get 0.7 for this query.
+
+In contrast, with Learning To Rank, search result ranks would be
+determined by a machine learning model without manually changing
+the boost of a field.
+
+With Learning To Rank we would go through the same process of
+rating the search results.
+
+```
+harry potter  0         Pottery barn
+harry potter  3         Harry Potter
+...
+```
+
+However, after doing this, we would train a model with this data
+to rank the results better. The model optimises for nDCG.
+
+We train the model using using a combination of relevancy judgements
+and features: numeric measures which indicate how well we think the
+result matches the query.
+
+The text similarity scores Elasticsearch assigns to fields in the
+document are examples of features. The model tries to find correlations
+between document features and relevancy judgements, which allows it to
+predict how to order documents we don't have relevancy judgements for.
+
+```
+query         rating    document      title_score view_count  recency  pagerank ...
+harry potter  0         Pottery barn  0.2         9876        0.12     9876     ...
+harry potter  3         Harry Potter  0.2         1234        1234     1234     ...
+...
+```
+
+For example, take the SVM dataset below for the query 'harry potter' (qid:1)
```

Comment:
SVM?

Reply:
The libsvm format (https://sourceforge.net/p/lemur/wiki/RankLib%20File%20Format/) which seems quite common in LTR datasets. We chose this because it's the format required by the TensorFlow ranking module.

Reply:
Ah, I see. Might be useful to mention that in the ADR as well - the acronym through me off a bit the first time I read it.

Reply:
Thanks, I added a note there :)